### PR TITLE
[alertmanager] Add support for PVC retention policy in statefulset

### LINKED ISSUE #3712: [alertmanager] StatefulSet PVC Retention Policy
### Is your feature request related to a problem ?

In Kubernetes 1.27 a new [feature](https://kubernetes.io/blog/2023/05/04/kubernetes-1-27-statefulset-pvc-auto-deletion-beta/#the-new-statefulset-pvc-retention-policy) was added where a retention policy can be used to control if and when PVCs created from a `StatefulSet`'s `volumeClaimTemplate` are deleted.

The alertmanager chart does not currently give users the ability to set `whenDeleted` or `whenScaled` inside the `persistentVolumeRetentionPolicy`.

### Describe the solution you'd like.

The ability to be able to set the alertmanager `persistentVolumeRetentionPolicy` `whenDeleted` or `whenScaled` to whatever I want using a boolean.

### Describe alternatives you've considered.

`NONE`

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by peterabarr**
#### What this PR does / why we need it

Kubernetes 1.27.0 enabled statefulset's [persistent volume claim retention policy](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention) (beta) by means of which a PVC can either be retained or deleted on deleting and scaling down the statefulset, respectively.

To enable the auto delete feature in the chart, this PR adds fields `Values.persistentVolumeClaimRetentionPolicy.whenDeleted` and `Values.persistentVolumeClaimRetentionPolicy.whenScaled`. Default value is `retain` (PVCs retained).

The values schema file has been updated with recently added fields.

#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #3712 

#### Checklist

- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

File: charts/alertmanager/values.yaml [hunk 1]
```diff
@@ -173,6 +173,15 @@ topologySpreadConstraints: []
...
   annotations: {}
 
+  ## Statefulset's persistent volume claim retention policy
+  ## pvcDeleteOnStsDelete and pvcDeleteOnStsScale determine whether
+  ## statefulset's PVCs are deleted (true) or retained (false) on scaling down
+  ## and deleting statefulset, respectively. Requires 1.27.0+.
+  ## Ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention
+  ##
+  pvcDeleteOnStsDelete: false
+  pvcDeleteOnStsScale: false
```
Comment (monotek) [inline charts/alertmanager/values.yaml:183]: can we have it like this please?
```suggestion
  persistentVolumeClaimRetentionPolicy:
    whenDeleted: Delete
    whenScaled: Delete
```

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/alertmanager/templates/statefulset.yaml [hunk 2]
```diff
@@ -11,6 +11,11 @@ metadata:
   {{- end }}
   namespace: {{ include "alertmanager.namespace" . }}
 spec:
+  {{- if semverCompare ">= 1.27.x" (include "alertmanager.kubeVersion" .) }}
+  persistentVolumeClaimRetentionPolicy:
+    whenDeleted: {{ ternary "Delete" "Retain" .Values.server.statefulSet.pvcDeleteOnStsDelete }}
```
Comment (monotek) [inline charts/alertmanager/templates/statefulset.yaml:16]: this needs to be adjusted too

File: charts/alertmanager/values.yaml [hunk 3]
```diff
@@ -235,6 +235,16 @@ topologySpreadConstraints: []
 statefulSet:
   annotations: {}
 
+  ## Statefulset's persistent volume claim retention policy
+  ## pvcDeleteOnStsDelete and pvcDeleteOnStsScale determine whether
+  ## statefulset's PVCs are deleted (true) or retained (false) on scaling down
+  ## and deleting statefulset, respectively. Requires 1.27.0+.
+  ## Ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention
+  ##
+persistentVolumeClaimRetentionPolicy:
```
Comment (monotek) [inline charts/alertmanager/values.yaml:244]: indent

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/alertmanager/values.yaml [hunk 4]
```diff
@@ -235,6 +235,17 @@ topologySpreadConstraints: []
...
 
+  ## Statefulset's persistent volume claim retention policy
+  ## pvcDeleteOnStsDelete and pvcDeleteOnStsScale determine whether
+  ## statefulset's PVCs are deleted (true) or retained (false) on scaling down
+  ## and deleting statefulset, respectively. Requires 1.27.0+.
+  ## Ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#persistentvolumeclaim-retention
+  ##
+  persistentVolumeClaimRetentionPolicy:
+    whenDeleted: Delete
+    whenScaled: Delete
```
Comment (monotek) [inline charts/alertmanager/values.yaml:246]: keep default
```suggestion
    whenDeleted: Retain
    whenScaled: Retain
```

File: charts/alertmanager/templates/statefulset.yaml [hunk 5]
```diff
@@ -11,6 +11,11 @@ metadata:
   {{- end }}
   namespace: {{ include "alertmanager.namespace" . }}
 spec:
+  {{- if semverCompare ">= 1.27.x" (include "alertmanager.kubeVersion" .) }}
+  persistentVolumeClaimRetentionPolicy:
+    whenDeleted: {{ ternary "Delete" "Retain" .Values.persistentVolumeClaimRetentionPolicy.whenDeleted }}
+    whenScaled: {{ ternary "Delete" "Retain" .Values.persistentVolumeClaimRetentionPolicy.whenScaled }}
```
Comment (monotek) [inline charts/alertmanager/templates/statefulset.yaml:17]:
```suggestion
    whenDeleted: {{ .Values.persistentVolumeClaimRetentionPolicy.whenDeleted }}
    whenScaled: {{ .Values.persistentVolumeClaimRetentionPolicy.whenScaled }}
```

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/alertmanager/Chart.yaml [hunk 6]
```diff
@@ -6,7 +6,7 @@ icon: https://raw.githubusercontent.com/prometheus/prometheus.github.io/master/a
 sources:
   - https://github.com/prometheus/alertmanager
 type: application
-version: 1.7.0
+version: 1.7.1
```
Comment (monotek) [inline charts/alertmanager/Chart.yaml:9]:
```suggestion
version: 1.8.0
```

**Review by monotek — CHANGES_REQUESTED** [review]

**Review by monotek — APPROVED** [review]

File: charts/alertmanager/values.schema.json [hunk 7]
```diff
@@ -912,4 +953,3 @@
             }
         }
     }
-}
```
Comment (monotek) [inline charts/alertmanager/values.schema.json:915]: this seems wrong

**Review by monotek — CHANGES_REQUESTED** [review]

File: charts/alertmanager/values.schema.json (hunk 7 above)
Comment (peterabarr) [reply charts/alertmanager/values.schema.json:915]: @monotek Thanks for taking a look at this PR. My linter is saying that the JSON is valid, would you mind checking again?

Comment (peterabarr) [reply charts/alertmanager/values.schema.json:915]: @monotek I have made changes to this file if you could re-review.

**Review by monotek — APPROVED** [review]

**Comment by peterabarr** [comment]
@monotek I had to make a change to fix the lint check failing & found a problem when helm templating in the statefulset.yaml file which I corrected if you could re-review, thanks

**Comment by VocalPalsy** [comment]
Thanks @peterabarr @monotek any updates on this we are waiting on this for our environment?

**Comment by monotek** [comment]
please fix ci error

**Review by monotek — CHANGES_REQUESTED** [review]
please fix ci error

**Pull request closed without merging**